opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -371,13 +371,19 @@ def _create_immutable_attributes(attributes):
     return MappingProxyType(attributes.copy() if attributes else {})
 
 
+class _SpanEndedException(Exception):
+    pass
+
+
 def _check_span_ended(func):
     def wrapper(self, *args, **kwargs):
-        with self._lock:  # pylint: disable=protected-access
-            if self.end_time is not None:
-                logger.warning("Calling %s on an ended span.", func.__name__)
-                return
-            func(self, *args, **kwargs)
+        try:
```

Comment:
I might consider re-writing this try/catch as an if/else instead. A good way to do this is to set a value inside the lock, and use that to determine if the logging call is needed outside.

Exception handling has a lot of overhead aside from branching.

Reply:
Good point, this just seemed like the cleanest way to implement. Have reworked to use if/else, please take a look